internal/util/util.go

```diff
@@ -0,0 +1,98 @@
+// Package util provides helper funtions to prepare materials for talking to attestation verifiers.
+package util
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/google/go-tpm-tools/cel"
+	"github.com/google/go-tpm-tools/client"
+	attestpb "github.com/google/go-tpm-tools/proto/attest"
+	"github.com/google/go-tpm-tools/verifier"
+	"github.com/google/go-tpm-tools/verifier/rest"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+)
+
+// TpmKeyFetcher abstracts the fetching of various types of Attestation Key from TPM
+type TpmKeyFetcher func(rw io.ReadWriter) (*client.Key, error)
+
+// PrincipalFetcher fetch ID token with specific audience from Metadata server.
+// See https://cloud.google.com/functions/docs/securing/authenticating#functions-bearer-token-example-go.
+func PrincipalFetcher(audience string, mdsClient *metadata.Client) ([][]byte, error) {
+	u := url.URL{
+		Path: "instance/service-accounts/default/identity",
+		RawQuery: url.Values{
+			"audience": {audience},
+			"format":   {"full"},
+		}.Encode(),
+	}
+	idToken, err := mdsClient.Get(u.String())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get principal tokens: %w", err)
+	}
+
+	tokens := [][]byte{[]byte(idToken)}
+	return tokens, nil
+}
+
+// GetAttestation gathers the materials required for remote attestation from TPM
+func GetAttestation(tpm io.ReadWriteCloser, akFetcher TpmKeyFetcher, nonce []byte) (*attestpb.Attestation, error) {
```

Comment:
I understand this is an open source project but I'd still suggest following Go Style from go/go-style/decisions#getters. Consider renaming it to something like `GenerateAttestation`.